gravity.sh

```diff
@@ -83,6 +89,74 @@ if [[ -r "${piholeDir}/pihole.conf" ]]; then
   echo -e "  ${COL_LIGHT_RED}Ignoring overrides specified within pihole.conf! ${COL_NC}"
 fi
 
+# Generate new sqlite3 file from schema template
+generate_gravity_database() {
+  sqlite3 "${gravityDBfile}" < "${gravityDBschema}"
+  chown $PIHOLE_USER:$PIHOLE_GROUP "${gravityDBfile}"
+}
+
+# Import domains from file and store them in the specified database table
+gravity_store_in_database() {
+  # Define locals
+  local table="${1}"
+  local source="${2}"
+  local template="${3}"
+
+  # Create database file if not present
+  if [ ! -e "${gravityDBfile}" ]; then
+    generate_gravity_database
+  fi
+
+  # Empty domains
+  output=$( { sqlite3 "${gravityDBfile}" <<< "DELETE FROM ${table};"; } 2>&1 )
```

Comment:
This makes the source of data the `source` file. If we are moving this data to the database, does it make sense to do this?

Reply:
I don't understand what you mean... There is no `source` in this line.

Reply:
This PR moves the storage of domains to the database, so we should not use the files to determine the domains to store.

Currently, the database domains just mirror the domains in the files, instead, the database should be the source of truth.

In order to do this, we will have to update all of the places which read, insert, or update the lists (`list.sh` and the web interface). We will have to decide if we want to update the PHP web interface or start using the database with the new web interface and API.

Reply:
> Currently, the database domains just mirror the domains in the files, instead, the database should be the source of truth.

The database is already the source of truth. We create a file containing all domains and mass-inject the domains into the database as this is faster than injecting each domain individually.

The corresponding branch of FTL cannot read the list files (it doesn't understand the simple list format we're using any longer) but sources the domains solely from the database file.

Reply:
It is the source of truth for FTL, but it itself is generated from the list files. In the code that is referenced by this thread, gravity deletes all of the entries from the table and uses the file to fill in the entries. That means that the database entries are sourced from the files. Instead, the database should just do a one-time import from the files (maybe just on upgrade). All future list operations should operate on the database instead of the files.

Reply:
This function is now only called:
- once for each file during upgrade to Pi-hole v5.0
- to move the generated `preEventHorizon` into the `gravity` table

The source files are always deleted after moving their content to the database.